Impossible-Agent-Suite/flash-loan/src/agent.ts

```diff
@@ -0,0 +1,112 @@
+import {
+  Finding,
+  HandleTransaction,
+  TransactionEvent,
+  FindingSeverity,
+  FindingType,
+  getEthersProvider,
+} from 'forta-agent';
+
+const { ethers } = require('ethers');
+
+import LRU from 'lru-cache';
+
+const cache: LRU<string, string> = new LRU<string, string>({ max: 10000 });
+
+let SWAP_FACTORY_ADDRESS = '0x918d7e714243f7d9d463c37e106235dcde294ffc';
+
+export const SWAP_FACTORY_ABI = [
+  'event PairCreated(address indexed token0, address indexed token1, address pair, uint)',
+  'function getPair(address tokenA, address tokenB) external view returns (address pair)',
+];
+
+export const PAIR_SWAP_ABI = [
+  'event Swap(address indexed sender, uint amount0In, uint amount1In, uint amount0Out, uint amount1Out, address indexed to)',
+  'function swap(uint amount0Out, uint amount1Out, address to, bytes calldata data)',
+  'function factory() external view returns (address)',
+  'function token0() external view returns (address)',
+  'function token1() external view returns (address)',
+];
+
+// Receives the address of a contract with same ABI as PAIR_SWAP_ABI and returns the factory address
+export const getContractFactory = async (swapContract: string) => {
+  let factory;
+  // If the cache contains the factory for `swapContract`
+  if(cache.has(swapContract)) {
+    // Set factory to the `swapContract` factory
+    factory = cache.get(swapContract);
+  } else {
+  // Otherwise if the cache does not contain the factory for `swapContract`
+    try {
+      // Query the smart contract for its factory
+      const provider = getEthersProvider();
+      const contractInterface = new ethers.Contract(swapContract, PAIR_SWAP_ABI, provider);
+      factory = await contractInterface.factory();
+      // Add it to the cache
+      cache.set(swapContract, factory);
+    } catch(error) {
+      // If the query fails (contract doesn't have a `factory` function then set to zero address
+      cache.set(swapContract, '0x0000000000000000000000000000000000000000');
+    }
```

Comment:
I think you should set `factory` variable here too, just to not return undefined

Reply:
Addressed in following commit.